### PR TITLE
New CNN model

### PR DESCRIPTION
**Pull request by jamesrequa**
- Implemented new CNN architecture
- Added BatchNormalization
- Added learning rate as a parameter in train_model method of
keras_helper and added as a user defined hyper-parameter in the
notebook
- Added methods for save_weights and load_weights to keras_helper;
functionality also working in the notebook
- Moved epochs to training block of notebook for easier access
- Added optional Continue Training block to notebook
- Fixed bugs with EarlyStopping and ModelCheckpoint

### CODE REVIEW

File: notebooks/amazon_forest_notebook.py [hunk 1]
```diff
@@ -223,17 +224,36 @@
...
+
+# Here you can continue training the model for additional epochs with the option to adjust the learning rate.
+# 
+# If you are finished training you can skip this step.
+
+# <codecell>
+
+epochs = 5
+learn_rate = 0.001
+train_losses, val_losses, fbeta_score = classifier.train_model(x_train, y_train, learn_rate, epochs, batch_size,
```
Comment (EKami) [inline notebooks/amazon_forest_notebook.py]: Here you are overriding the `train_losses` and `val_losses` arrays from the 1st training pass. Instead you should append the results to the previous ones or do 2 separate plots below

File: src/keras_helper.py [hunk 2]
```diff
@@ -31,51 +31,71 @@ def __init__(self):
 
     def add_conv_layer(self, img_size=(32, 32), img_channels=3):
         self.classifier.add(BatchNormalization(input_shape=(*img_size, img_channels)))
+
+        self.classifier.add(Conv2D(32, (3, 3), padding='same', activation='relu'))
         self.classifier.add(Conv2D(32, (3, 3), activation='relu'))
-        self.classifier.add(MaxPooling2D(pool_size=(2, 2)))
+        self.classifier.add(MaxPooling2D(pool_size=2))
```
Comment (godardt) [inline src/keras_helper.py:37]: This is actually the same thing but it's just a matter of preferences. I wanted to keep a tuple to not forget that it's the 2x2 patch but you can keep it like that if you want :)

File: notebooks/amazon_forest_notebook.py (hunk 1 above)
Comment (godardt) [inline notebooks/amazon_forest_notebook.py]: Here you are overriding the `train_losses` and `val_losses` arrays from the 1st training pass. Instead you should append the results to the previous ones or do 2 separate plots below

File: notebooks/amazon_forest_notebook.py [hunk 3]
```diff
@@ -208,7 +209,7 @@
 
 # <codecell>
 
-from keras.callbacks import ModelCheckpoint
+from tensorflow.contrib.keras.api.keras.callbacks import ModelCheckpoint
 
 filepath="weights.best.hdf5"
 checkpoint = ModelCheckpoint(filepath, monitor='val_acc', verbose=1, save_best_only=True, mode='max')
```
Comment (godardt) [inline notebooks/amazon_forest_notebook.py]: The `mode='max'` is not needed here. The default value is ok for our case :)

Comment (godardt) [inline notebooks/amazon_forest_notebook.py]: Are you sure we want to keep the `verbose=1`? It really generates ugly output and doesn't give us much useful information:
amazon_forest_notebook

Something like this would be best:
`checkpoint = ModelCheckpoint(filepath, monitor='val_acc', save_best_only=True)`
Wdyt?

**Review by godardt — CHANGES_REQUESTED** [review]
Great job!

File: notebooks/amazon_forest_notebook.py (hunk 3 above)
Comment (jamesrequa) [reply notebooks/amazon_forest_notebook.py]: Yes the output isn't pretty but its nice to see at a quick glance that the model has achieved a new best weight on a particular epoch. Otherwise you have to search for the accuracy numbers to see for yourself which one is the best.

Comment (godardt) [reply notebooks/amazon_forest_notebook.py]: Ok we'll keep it verbose then :)

File: src/keras_helper.py (hunk 2 above)
Comment (jamesrequa) [reply src/keras_helper.py:37]: Yea I prefer just having one integer since it defaults to equal pool dims, hope you don't mind! ha

Comment (godardt) [reply src/keras_helper.py:37]: Yes np :)

File: notebooks/amazon_forest_notebook.py (hunk 3 above)
Comment (jamesrequa) [reply notebooks/amazon_forest_notebook.py]: OK will remove it

File: notebooks/amazon_forest_notebook.py (hunk 1 above)
Comment (jamesrequa) [reply notebooks/amazon_forest_notebook.py]: Thanks for catching this. Revising to append results  for `train_losses` and `val_losses` each time so its not overwritten.

File: notebooks/amazon_forest_notebook.py [hunk 4]
```diff
@@ -212,39 +210,40 @@
...
-train_losses, val_losses, fbeta_score = classifier.train_model(x_train, y_train, learn_rate, epochs, batch_size, 
-                                                               validation_split_size=validation_split_size, 
-                                                               train_callbacks=[checkpoint])
 
-# <markdowncell>
-
-# ## Continue Training
+train_losses, val_losses = [], []
+epochs_arr = [20, 5, 5]
+learn_rates = [0.001, 0.0001, 0.00001]
```
Comment (godardt) [inline notebooks/amazon_forest_notebook.py]: We could move these parameters (epoch and learn_rates) in the `Define hyperparameters` cell or change/remove the `Define hyperparameters` cell. It doesn't make much sense to have hyperparameters spread in the notebook when there is 1 place where they should be defined :)

File: notebooks/amazon_forest_notebook.py [hunk 5]
```diff
@@ -212,39 +210,40 @@
...
 # <markdowncell>
 
-# ## Create the neural network definition
+# ## Define and Train model
+
+# <markdowncell>
+
+# Here we define the model and begin training. 
+# 
+# Before starting the training process, you should first set a learning rate annealing optimization schedule by choosing a series of learning rates (learn_rates) with corresponding number of epochs for each (epochs_arr).
```
Comment (godardt) [inline notebooks/amazon_forest_notebook.py]: In markdown code variables are usually contained in ` quotes.

...schedule by choosing a series of learning rates (`learn_rates`) with...

...number of epochs for each (`epochs_arr`).

**Review by godardt — CHANGES_REQUESTED** [review]

File: notebooks/amazon_forest_notebook.py [hunk 6]
```diff
@@ -208,32 +207,52 @@
...
+
+train_losses, val_losses = [], []
+epochs_arr = [20, 5, 5]
+learn_rates = [0.001, 0.0001, 0.00001]
+for learn_rate, epochs in zip(learn_rates, epochs_arr):
+    tmp_train_losses, tmp_val_losses, fbeta_score = classifier.train_model(x_train, y_train, learn_rate, epochs, 
+                                                                           batch_size, validation_split_size=validation_split_size, 
+                                                                           train_callbacks=[checkpoint])
+    train_losses.append(tmp_train_losses)
+    val_losses.append(tmp_val_losses)
```
Comment (godardt) [inline notebooks/amazon_forest_notebook.py]: This will append lists into lists, instead we want to append the value to the end of the original list so the correct behaviour here would be using `+=` instead of `append`

**Review by godardt — CHANGES_REQUESTED** [review]

File: notebooks/amazon_forest_notebook.py [hunk 7]
```diff
@@ -214,15 +210,32 @@
 
 # <markdowncell>
 
+# ## Choose Hyperparameters
+# 
+# Choose your hyperparameters below for training. 
+# 
+# If you'd like to create a learning rate annealing schedule you can set a series of learning rates with `lr_1, lr_2, lr_3` and corresponding number of epochs for each `epochs_1, epochs_2, epochs_3`. 
+# 
+# Alternatively, if you just want to run one training session at a fixed learning rate and num epochs you can just enter values for `lr_1` and `epochs_2` only.
```
Comment (godardt) [inline notebooks/amazon_forest_notebook.py]: Following these instructions I'll be tempted to define only the following learning rates and epochs:
```
lr_1 = 0.001
epochs_1 = 10
```
Except that if I do that the code below will crash :P

It's better to use directly the epoch and lr arrays and let the user add as much values as we want in it :)

**Review by godardt — CHANGES_REQUESTED** [review]

File: notebooks/amazon_forest_notebook.py (hunk 7 above)
Comment (jamesrequa) [reply notebooks/amazon_forest_notebook.py]: Oops I should have tested it first to see if that would work. OK.

Comment (jamesrequa) [reply notebooks/amazon_forest_notebook.py]: @EKami Changes already made, please update review!

File: notebooks/amazon_forest_notebook.py [hunk 8]
```diff
@@ -201,39 +198,72 @@
...
+checkpoint = ModelCheckpoint(filepath, monitor='val_acc', verbose=1, save_best_only=True)
 
 # <markdowncell>
 
-# Creating a checkpoint saves the best model weights across all epochs in the training process. This ensures that we will always use only the best weights when making our predictions on the test set rather than using the default which takes the final score from the last epoch. 
+# ## Choose Hyperparameters
+# 
+# Choose your hyperparameters below for training. 
+# 
+# We have assigned a series of learning rate values to create an optimized annealing schedule `lr_1, lr_2, lr_3` and corresponding number of epochs for each `epochs_1, epochs_2, epochs_3`. Feel free to change these values or you can just use the defaults we have filled in.
```
Comment (godardt) [inline notebooks/amazon_forest_notebook.py]: Hmm, sorry for not being clear but what I wanted to say is that defining variables which are just used to fill an array just adds a burden to the code 😅 .  For instance if I want to add another value to the array and follow this pattern I would have to add `lr_4` and `epochs_4` then add them to the array (where I could just add them to the array). It would be much easier to directly have the arrays (such as `epochs_arr = [0.001, 0.0001, 0.00001]`) and explain in the comment why we are creating such arrays. These kind of behaviours can often lead to what we call [edge cases](https://en.wikipedia.org/wiki/Edge_case) in the dev world.

**Review by godardt — CHANGES_REQUESTED** [review]

File: notebooks/amazon_forest_notebook.py [hunk 9]
```diff
@@ -213,27 +213,19 @@
...
 
 # ## Define and Train model
-
-# <markdowncell>
-
-# Here we define the model and begin training.
+# 
+# Here we define the model and begin training. 
+# 
+# Note that we have created a learning rate annealing schedule with a series of learning rates as defined `lr_1, lr_2, lr_3` and corresponding number of epochs for each `epochs_1, epochs_2, epochs_3`. Feel free to change these values if you like or just use the defaults.
```
Comment (godardt) [inline notebooks/amazon_forest_notebook.py]: > we have created a learning rate annealing schedule with a series of learning rates as defined `lr_1, lr_2, lr_3` and corresponding number of epochs for each `epochs_1, epochs_2, epochs_3`

This message is outdated x)

**Review by godardt — CHANGES_REQUESTED** [review]

File: notebooks/amazon_forest_notebook.py (hunk 9 above)
Comment (jamesrequa) [reply notebooks/amazon_forest_notebook.py]: @EKami Sorry about that...fixed

**Review by godardt — APPROVED** [review]

**Comment by jamesrequa** [comment]
[BatchNorm](https://trello.com/c/6yxxWsW7/2-batchnorm)

**Comment by jamesrequa** [comment]
[Learning Rate Annealing/Simulated annealing](https://trello.com/c/Hjs2DbJV/11-learning-rate-annealing-simulated-annealing)
